graphene/src/document.rs

```diff
@@ -92,6 +95,52 @@ impl Document {
 		self.folder_mut(path)?.layer_mut(id).ok_or_else(|| DocumentError::LayerNotFound(path.into()))
 	}
 
+	fn shape_as_seen(&self, path: &[LayerId]) -> Result<Shape, DocumentError> {
+		fn helper(layer: &Layer, mut transform: DAffine2, mut slice: usize, path: &[LayerId]) -> Result<Shape, DocumentError> {
+			match &layer.data {
+				LayerDataType::Shape(shape) => {
+					let mut clone = shape.clone();
+					transform = layer.transform * transform;
+					clone.path.apply_affine(Affine::new(transform.to_cols_array()));
+					Ok(clone)
+				}
+				LayerDataType::Folder(folder) => {
+					transform = layer.transform * transform;
+					if slice >= path.len() {
+						return Err(DocumentError::InvalidPath);
+					}
+					if let Some(sub_layer) = folder.layer(path[slice]) {
+						slice += 1;
+						return helper(sub_layer, transform, slice, path);
+					}
+					Err(DocumentError::InvalidPath)
+				}
+			}
+		}
+		// skip the root layer, whose transform doesn't effect the coordinates
+		if let LayerDataType::Folder(root) = &self.root.data {
+			match root.layer(path[0]) {
+				Some(layer) => helper(layer, DAffine2::IDENTITY, 1, path),
+				None => Err(DocumentError::InvalidPath),
+			}
+		} else {
+			Err(DocumentError::NotAFolder)
+		}
+	}
+
+	/// Return vector of immutable references to each shape specified in paths
+	/// If any path is not a shape, or does not exist, DocumentError::InvalidPath is returned
+	fn shapes_as_seen(&self, paths: &Vec<Vec<LayerId>>) -> Result<Vec<Shape>, DocumentError> {
```

Comment:
Perhaps `projected_shapes`

Reply:
`shapes_as_seen` definitely isn't a great name. I'm thinking maybe `transformed_shapes`, because typically a projection is onto something

Reply:
It is projecting the shapes into viewport space, however `transformed_shapes` is also a good name.